### DISCUSSION TITLE
AsyncResponseStream is not streaming

### DISCUSSION THREAD

**Discussion by y3nd** [mathieucarbou/ESPAsyncWebServer#163] [Q&A]
Hello,
I would like to use AsyncResponseStream similarly to what I do with a Serial stream. I need to send several 10th of kilobytes in binary and I have heap size issue.
Since a bufferSize is specified, I had initially understood that it would push data to the client progressively and maybe block the writes to slow down like Serial do (i think).
I then noticed that the content-length was set on the response and since headers should be sent before the body, I understood that it was entirely buffered and not sent like a stream.
I then used .availableForWrite to prevent filling the buffer but it seems not implemented.

From user POV, what is the purpose of [bufferSize here](https://github.com/mathieucarbou/ESPAsyncWebServer/blob/bccee27d2ba4dbba39a670274ad4b4fcd537ece5/src/WebResponseImpl.h#L148) and how could send the response progressively, maybe by specifying the content length before and using another async response type ?

Thanks

**Comment by mathieucarbou** [comment]
Hello,

There is an example about serving a big html fie using a small buffer and with caching headers here:

https://github.com/mathieucarbou/ESPAsyncWebServer/blob/main/examples/SimpleServer/SimpleServer.ino#L472-L504

**Comment by mathieucarbou** [comment]
`AsyncResponseStream` is badly named: it is not streaming: it is just implementing the `Print` interface and collects the data into the buffer, allocated with `bufferSize` bytes.

If it was really streaming, its API should have been like that:

`AsyncResponseStream(const char* contentType, Stream& stream, size_t chunkBufferSize);`

and the callback as seen in the link above would then read the stream and fill the chunk buffer.

Note: this is nearly what `AsyncStreamResponse` does.

You have:
- `AsyncStreamResponse`
- `AsyncResponseStream`

`AsyncWebServerResponse* AsyncWebServerRequest::beginResponse(Stream& stream, const char* contentType, size_t len, AwsTemplateProcessor callback) {`

very confusing... ;-)
